les/balance.go

```diff
@@ -17,24 +17,101 @@
 package les
 
 import (
+	"math"
 	"sync"
 	"time"
 
 	"github.com/ethereum/go-ethereum/common/mclock"
 )
 
+const maxBalance = math.MaxInt64
+
 const (
 	balanceCallbackQueue = iota
 	balanceCallbackZero
 	balanceCallbackCount
 )
 
+// expirationController controls the exponential expiration of positive and negative
+// balances
+type expirationController interface {
+	posExpiration(mclock.AbsTime) uint64
+	negExpiration(mclock.AbsTime) uint64
+}
+
+// expiredValue is a scalar value that is continuously expired (decreased exponentially)
+// based on the logarithmic expiration offset value provided by the expirationController.
+// The actual value can be calculated as base*2^(exp-logOffset/log2Multiplier).
+//
+// Note: this representation is basically a floating point value capable of handling
+// 64 bit exponents. The operations are not general purpose floating point operations
+// though because a monotonously increasing exponent is assumed and therefore if an
+// operation is performed with two different exponents then simply the higher one is used.
+type expiredValue struct {
+	base, exp uint64
+}
+
+// value calculates the value at the given moment
+func (e expiredValue) value(logOffset uint64) uint64 {
+	return uint64(math.Exp(float64(int64(e.exp*log2Multiplier-logOffset))/logMultiplier) * float64(e.base))
+}
+
+// add adds a signed value at the given moment
```

Comment:
@rjl493456442 here is a commented version of the function (will add it later to the code if you say it explains it better)
```
// add adds a signed value at the given moment
func (e *expiredValue) add(amount int64, logOffset uint64) int64 {
	// amount is exponentially scaled up based on logOffset to match the scale of value e
	// new exponent is directly based on the higher 40 bits of logOffset (a 2^24 increase in logOffset means a
	// 2x upscaling of expiredValue which can be expressed as a +1 increase in the base 2 exponent)
    addExp := logOffset / log2Multiplier
	// the scale factor for the base value (amount) is based on the lower 24 bits of logOffset (1 <= baseMul < 2)
    baseMul := math.Exp(float64(logOffset%log2Multiplier) / logMultiplier)
    if addExp < e.exp {
		// this can only happen if the value is added with an earlier logOffset than the last one
		// do not lower e.exp in this case, lower the scale of the added value instead
        baseMul /= math.Pow(2, float64(e.exp-addExp))
    }
    if addExp > e.exp {
		// new value has a higher exponent, use the newer exponent from now on
		// divide the old base value by a power of 2 (bitwise right shift)
        e.base >>= (addExp - e.exp)
        e.exp = addExp
    }
	// add the amount and protect against going below 0
    add := int64(float64(amount) * baseMul)
    if add >= 0 || uint64(-add) <= e.base {
        e.base += uint64(add)
        return amount
    } else {
        e.base = 0
        return int64(-float64(e.base) / baseMul)
    }
}
```